### PR TITLE
feat: full text search for `TeamCollections` and `TeamRequests`

### PR DESCRIPTION
**Pull request by balub**
Closes HSB-322

### Description

In this PR we introduce a new route that allows for full-text search to be performed over saved team collections and requests. 

`GET http://localhost:3170/v1/team-collection/search/:teamID/:searchQuery?take=5&skip=2`

We also have simple offset pagination available using the `take` and `skip` query params.

An example response to the above query may be like the following:

```
{
  "data": [
   .........,
    {
      "type": "collection",
      "title": "Get Started for BSPs",
      "id": "clt1k7d6u02h9ioerkkia6b7h",
      "path": [
        {
          "id": "clt1k7d6p02goioeri3f85nju",
          "title": "Cloud API",
          "type": "collection",
          "path": []
        }
      ]
    },
   .........,
    {
      "type": "request",
      "title": "get request",
      "method": "GET",
      "id": "clt1k7cm9000hioerpzpfpmb0",
      "path": [
        {
          "id": "clt1k7cm9000fioer57bl0n9d",
          "title": "child collection",
          "type": "collection",
          "path": {
            "id": "clt1k7cm70006ioer65sc2xsb",
            "title": "Currency Rates API",
            "type": "collection",
            "path": []
          }
        }
      ]
    },
   .........,
  ]
}
```

### Checks

- [x] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [x] All the tests have passed

### CODE REVIEW

File: packages/hoppscotch-backend/src/team-collection/team-collection.service.ts [hunk 1]
```diff
@@ -1056,4 +1064,232 @@ export class TeamCollectionService {
...
+    }
+  }
+
+  /**
+   * Generate the parent tree from the collections
+   *
+   * @param parentCollections The parent collections
+   * @returns The parent tree of the parent collections
+   */
+  private generateParentTree(parentCollections) {
```
Comment (mirarifhasan) [inline packages/hoppscotch-backend/src/team-collection/team-collection.service.ts:1221]: Can we add a type for `parentCollections` parameter?

File: packages/hoppscotch-backend/src/team-collection/team-collection.service.ts [hunk 2]
```diff
@@ -1056,4 +1064,232 @@ export class TeamCollectionService {
...
+    return null;
+  }
+
+  /**
+   * Generate the parent tree of a request
+   *
+   * @param id The ID of the request
+   * @returns The parent tree of the request
+   */
+  private async fetchRequestParentTree(id) {
```
Comment (mirarifhasan) [inline packages/hoppscotch-backend/src/team-collection/team-collection.service.ts:1269]:
```suggestion
  private async fetchRequestParentTree(id: string) {
```

File: packages/hoppscotch-backend/src/team-collection/team-collection.controller.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import { throwHTTPErr } from 'src/auth/helper';
+import * as E from 'fp-ts/Either';
+import { ThrottlerBehindProxyGuard } from 'src/guards/throttler-behind-proxy.guard';
+import { JwtAuthGuard } from 'src/auth/guards/jwt-auth.guard';
+import { RequiresTeamRole } from 'src/team/decorators/requires-team-role.decorator';
+import { TeamMemberRole } from '@prisma/client';
+import { RESTTeamMemberGuard } from 'src/team/guards/rest-team-member.guard';
+import { take } from 'rxjs';
+
+@UseGuards(ThrottlerBehindProxyGuard)
```
[the remark this answers is not in the document]
Comment (balub) [reply packages/hoppscotch-backend/src/team-collection/team-collection.controller.ts:12]: I think we can leave it as it is for now, if we start noticing this thing causing any issues we can always remove/modify it.

File: packages/hoppscotch-backend/src/team-collection/team-collection.service.ts [hunk 4]
```diff
@@ -1056,4 +1063,248 @@ export class TeamCollectionService {
...
+    skip: number,
+  ) {
+    const query = Prisma.sql`
+    select id,title,'collection' AS type
+    from "TeamCollection"
+    where titlesearch @@ to_tsquery(${searchQuery})
+    order by ts_rank(titlesearch,to_tsquery(${searchQuery}))
+    limit ${take}
+    OFFSET ${skip === 0 ? 0 : (skip - 1) * take};
+  `;
```
Comment (mirarifhasan) [inline packages/hoppscotch-backend/src/team-collection/team-collection.service.ts:1143]: In this query, we missed to add another where condition on `teamID`.

The consequence is,
Assume, there are 2 teams. One is `Balu's Team`, another `Arif's Team`
balu@hopp.io is the only member of `Balu's Team`. arif@hopp.io is the only member of `Arif's Team`.

Now, when Balu goes to search, he gets suggestions/responses from 'Arif's Team' as well.

Comment (balub) [reply packages/hoppscotch-backend/src/team-collection/team-collection.service.ts:1143]: Understood, I'll push a fix for that ASAP.

**Review by mirarifhasan — APPROVED** [review]
LGTM

**Pull request merged**